include/openrave/interface.h

```diff
@@ -75,13 +75,67 @@ class OPENRAVE_API InfoBase
     virtual void DeserializeJSON(const rapidjson::Value& value, dReal fUnitScale, int options) = 0;
 };
 
+struct OPENRAVE_API ReadableInterfaceBase {
+    virtual ~ReadableInterfaceBase() = default;
+
+    typedef std::map<std::string, ReadablePtr, CaseInsensitiveCompare> READERSMAP;
+
+    /// \brief Returns the raw map reference, this is \b not multithread safe and the GetReadableInterfaceMutex should be locked before using.
+    inline const READERSMAP& GetReadableInterfaces() const
+    {
+        return __mapReadableInterfaces;
+    }
+
+    /// \brief Returns the readable interface. <b>[multi-thread safe]</b>
+    virtual ReadablePtr GetReadableInterface(const std::string& id) const;
+
+    /// \brief Set a new readable interface and return the previously set interface if it exists. <b>[multi-thread safe]</b>
+    virtual ReadablePtr SetReadableInterface(const std::string& id, ReadablePtr readable);
+
+    /// \brief sets a set of readable interfaces all at once. The pointers are copied
+    ///
+    /// \param mapReadables the readable interfaces to ste
+    /// \param bClearAllExisting if true, then clears the existing readables, if false, just updates the readables that are specified in mapReadables
+    virtual void SetReadableInterfaces(const READERSMAP& mapReadables, bool bClearAllExisting);
+
+    /// \brief clears the readable interfaces
+    virtual void ClearReadableInterfaces();
+    virtual void ClearReadableInterface(const std::string& id);
+
+    /// \brief updates the readable interfaces. returns true if there are any changes
+    virtual bool UpdateReadableInterfaces(const std::map<std::string, ReadablePtr>& newReadableInterfaces);
+
+    boost::shared_mutex& GetReadableInterfaceMutex() const
```

Comment:
@superfashi You added this mutex, any reason why it wasn't needed before?

Reply:
Isn't locking the env itself enough?

Reply:
It was there before inside `InterfaceBase`, but everything was using one mutex lock. I think locking the env is fine but is it better to have finer lock granularity?

Reply:
I see. There was a lock already.